external-crates/move/crates/move-analyzer/src/inlay_hints.rs

```diff
@@ -44,44 +46,112 @@ fn inlay_hints(context: &Context, fpath: PathBuf) -> Option<Vec<InlayHint>> {
     let symbols_map = &context.symbols.lock().ok()?;
     let symbols =
         SymbolicatorRunner::root_dir(&fpath).and_then(|pkg_path| symbols_map.get(&pkg_path))?;
-    inlay_hints_internal(symbols, fpath)
+    inlay_hints_internal(
+        symbols,
+        fpath,
+        context.inlay_type_hints,
+        context.inlay_param_hints,
+    )
 }
 
-fn inlay_hints_internal(symbols: &Symbols, fpath: PathBuf) -> Option<Vec<InlayHint>> {
+fn inlay_type_hints_internal(symbols: &Symbols, mod_defs: &ModuleDefs, hints: &mut Vec<InlayHint>) {
+    for untyped_def_loc in mod_defs.untyped_defs() {
+        let start_position = symbols.files.start_position(untyped_def_loc);
+        if let Some(DefInfo::Local(n, t, _, _, _)) = symbols.def_info(untyped_def_loc) {
+            let position = Position {
+                line: start_position.line_offset() as u32,
+                character: start_position.column_offset() as u32 + n.len() as u32,
+            };
+            let colon_label = InlayHintLabelPart {
+                value: ": ".to_string(),
+                tooltip: None,
+                location: None,
+                command: None,
+            };
+            let type_label = InlayHintLabelPart {
+                value: type_to_ide_string(t, /* verbose */ true),
+                tooltip: None,
+                location: None,
+                command: None,
+            };
+            let h = InlayHint {
+                position,
+                label: InlayHintLabel::LabelParts(vec![colon_label, type_label]),
+                kind: Some(InlayHintKind::TYPE),
+                text_edits: None,
+                tooltip: additional_hint_info(t, symbols),
+                padding_left: None,
+                padding_right: None,
+                data: None,
+            };
+            hints.push(h);
+        }
+    }
+}
+
+fn inlay_param_hints_internal(
+    symbols: &Symbols,
+    mod_defs: &ModuleDefs,
+    hints: &mut Vec<InlayHint>,
+) {
+    for call_info in mod_defs.call_infos.values() {
+        let Some(def_loc) = call_info.def_loc else {
+            continue;
+        };
+        let Some(DefInfo::Function(.., args, _, _, _)) = symbols.def_info.get(&def_loc) else {
+            continue;
+        };
+        if call_info.dot_call && args.is_empty() {
+            // methods should have at least one argument
+            continue;
+        };
+        for (name, loc) in args
+            .iter()
+            .skip(if call_info.dot_call { 1 } else { 0 })
+            .zip(&call_info.arg_locs)
+        {
+            let name_label = InlayHintLabelPart {
+                value: name.to_string(),
+                tooltip: None,
```

Comment:
More for my curiosity: would populating this field mean that the inlayed param name can _also_ have a tooltip (and, we could, e.g., put the type hint in this tooltip?). Not suggesting we do that, just wondering.

Reply:
Indeed. It's supported for type inlay hints and I was thinking of supporting it for param inlay hints but did not immediately see a significant benefit. You are right, though, it is useful to see the type of parameter when hovering over the hint and so it is in now! 

We can't do go-to-def/typedef at the moment due to what I think is a VSCode bug (https://github.com/microsoft/vscode/issues/213330)

Reply:
Interesting, that's cool! I mostly rely on hovers without inlays, but if I did use inlays having the type hover would basically be a no-brainer benefit.